Experimental Single Abstract Method support (sammy meets world)

adriaanm opened the pull request:
Under `-Xexperimental`, `typedFunction` invokes `synthesizeSAMFunction` when the expected type for the function literal (`pt`) is not the built-in `FunctionN` type of the expected arity, but `pt` does have a SAM with the expected number of arguments.

`synthesizeSAMFunction` expands the following tree:

```
{ (p1: T1, ..., pN: TN) => body } : S
```

to:

```
{
 def apply$body(p1: T1, ..., pN: TN): T = body

 new S { def apply(p1: T1, ..., pN: TN): T = apply$body(p1,..., pN) }
}
```

The expansion assumes `S` (the expected type) defines a single abstract method (let's call that method `apply` for simplicity).
1. If `T` is not fully defined, it is inferred by type checking `def apply$body` without a result type before type checking the block. The method's inferred result type is used instead of `T`. [See test case](https://github.com/adriaanm/scala/blob/67062db57c1abef88e0049dac5d82d4f13375a48/test/files/pos/sammy_poly.scala)
2. To more easily enforce that `S`'s members are not in scope in `body`, that tree goes to the `apply$body` method that's outside the anonymous subclass of S. (The separate `apply$body` method simplifies the implementation of 1&2.) [See test case](https://github.com/adriaanm/scala/blob/67062db57c1abef88e0049dac5d82d4f13375a48/test/files/pos/sammy_scope.scala)
3. The following [restrictions](https://github.com/adriaanm/scala/commit/657e85fe2412cdadc5ee9dc348159b32dcdfcba7#diff-2b1bf15b53df94495f7155a69f822adeR727) apply to `S`:
   1. Its primary constructor (if any) must be public, no-args, not overloaded.
   2. S must have exactly one abstract member, its SAM
   3. SAM must take exactly one argument list
   4. SAM must be monomorphic

We may later relax these requirements to allow an implicit argument list, both on the constructor and the SAM. Could also let the SAM be polymorphic.

PS: We'll require `import language.sam` instead of `-Xexperimental`, as soon as the SIP is ready and there are more tests.

adriaanm committed:
Single Abstract Method support: synthesis helpers

`synthesizeSAMFunction` will be used to expand the following tree:

```
{ (p1: T1, ..., pN: TN) => body } : S
```

to:

```
{
 def apply$body(p1: T1, ..., pN: TN): T = body

 new S { def apply(p1: T1, ..., pN: TN): T = apply$body(p1,..., pN) }
}
```

The expansion assumes `S` (the expected type) defines a single abstract method
(let's call that method `apply` for simplicity).

1. If 'T' is not fully defined, it is inferred by type checking
`def apply$body` without a result type before type checking the block.
The method's inferred result type is used instead of T`.
[See test/files/pos/sammy_poly.scala]

2. To more easily enforce S's members are not in scope in `body`, that tree
goes to the `apply$body` method that's outside the anonymous subclass of S.
(The separate `apply$body` method simplifies the implementation of 1&2.)

3. The following restrictions apply to S:
  1. Its primary constructor (if any) must be public, no-args, not overloaded.
  2. S must have exactly one abstract member, its SAM
  3. SAM must take exactly one argument list
  4. SAM must be monomorphic

We may later relax these requirements to allow an implicit argument list,
both on the constructor and the SAM. Could also let the SAM be polymorphic.

adriaanm committed:
Single Abstract Method support: synthesize SAMs

Under `-Xexperimental`, `typedFunction` invokes `synthesizeSAMFunction`
when the expected type for the function literal (`pt`) is not the built-in
`FunctionN` type of the expected arity, but `pt` does have a SAM
with the expected number of arguments.

PS: We'll require `import language.sam` instead of `-Xexperimental`,
as soon as the SIP is ready and there are more tests.

adriaanm committed:
Simplify partest.task target, fix typo in comment.

adriaanm committed:
Clarify findMembers, add reverse engineered docs

When looking for deferred members, it only makes sense
to retry when deferred members aren't excluded.

adriaanm commented:
Review by everyone (@gkossakowski, @jamesiry, @paulp, @retronym)

/cc @Alefas, @heathermiller, @xeno-by

adriaanm commented:
PS: I also [stress tested sammy](https://github.com/adriaanm/scala/commit/f29f76e)

milessabin commented:
Any chance of you dropping restrictions (3) and (4) ... especially (4).

paulp commented:
The line following (3) and (4) says "We may later relax these requirements to allow an implicit argument list, both on the constructor and the SAM. Could also let the SAM be polymorphic." The facts that the restrictions exist and that the questions were answered preemptively both serve to suggest that the restriction won't be dropped in the immediate future, which is what you're presumably asking if you got that far, though it sounds more like you didn't see that line.

milessabin commented:
@paulp I didn't see that line ;-)

retronym commented:
Repeated params should probably be ruled out early, currently we hit a type error:

``` scala
scala> trait T { def ap(as: String*): Any }; (xs => ()): T
<console>:10: error: type mismatch;
 found   : Seq[String]
 required: String
        (xs => ()): T
            ^
```

retronym commented:
By name looks to work, here are a few test cases:

```
scala> trait T { def ap(as: => String): Any }; ((x => ()): T).ap(???)
defined trait T
res9: Any = ()

scala> trait T { def ap(as: => String): Any }; ((x => x): T).ap(???)
scala.NotImplementedError: an implementation is missing
  at scala.Predef$.$qmark$qmark$qmark(Predef.scala:229)
```

retronym commented on src/compiler/scala/tools/nsc/typechecker/Typers.scala:
```diff
@@ -2693,17 +2693,198 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+          Nil,
+          List(fun.vparams.map(_.duplicate)), // must duplicate as we're also using them for `samDef`
+          TypeTree(samDefTp) setPos sampos.focus,
+          fun.body)
```
We're dropping prefixes in the signature of the SAM.

``` scala
% cat sandbox/test.scala
object Test {
  trait T { type U; def ap(a: String): U }; (xs => ???): T
}

% qbin/scalac -Xprint:typer -Xexperimental sandbox/test.scala
[[syntax trees at end of                     typer]] // test.scala
package <empty> {
  object Test extends scala.AnyRef {
    def <init>(): Test.type = {
      Test.super.<init>();
      ()
    };
    abstract trait T extends scala.AnyRef {
      type U;
      def ap(a: String): T.this.U
    };
    ({
      def ap$body(xs: String): Test.T#U = scala.this.Predef.???;
      @SerialVersionUID(0) final class $anonfun extends AnyRef with Test.T with Serializable {
        def <init>(): <$anon: Test.T> = {
          $anonfun.super.<init>();
          ()
        };
        final override def ap(xs: String): Test.T#U = ap$body(xs)
      };
      new $anonfun()
    }: Test.T)
  }
}

sandbox/test.scala:2: error: overriding method ap in trait T of type (a: String)this.U;
 method ap has incompatible type
  trait T { type U; def ap(a: String): U }; (xs => ???): T
                                                ^
```

Elsewhere, we call `pt memberInfo sam` which gives `(a: String)Test.T#U`. Instead, we would need: `pt.typeSymbol.thisType memberInfo sam` which gives `(a: String)T.this.U`. This in turn presents a challenge: we still need to `substituteThis` with the symbol of the derived class, which we're trying to avoid creating a symbol for manually.

adriaanm replied:
Yeah, this is a tricky problem. I hadn't thought of this instance, but a similar one applies when the result type depends on an argument value. The symbols for the params of the $body method will appear in the regular method's result type as we just copy it over.

adriaanm replied:
I can't think of a sensible implementation other than `???` for the lambda, though.

adriaanm committed:
Extract SerialVersionUIDAnnotation. Make SAM body synthetic.

Addressing review feedback.

retronym committed:
Don't pursue SAM translation after an arity mismatch.

Before this change:

    scala> trait T { def apply(a: Int): Int }
    defined trait T

    scala> ((x: Int, y: Int) => 0): T
    <console>:9: error: object creation impossible, since method apply in trait T of type (a: Int)Int is not defined
                  ((x: Int, y: Int) => 0): T
                                    ^
After the change, these cases report the same errors as they do
*without* -Xexperimental.

retronym commented:
Here a tweak to avoid pursuing SAM translation when arities mismatch:

https://github.com/adriaanm/scala/pull/8

adriaanm commented:
Thanks! I'm working on implementing the rest of your feedback here: https://github.com/adriaanm/scala/tree/sammy-wip -- last TODO is the java8 test.

retronym commented:
One area for refinement after this PR is error messages. They usually suffer when typechecking cascades through a fallback.

We should endeavour to:

``` scala
if (expectedTypeIsSam)
  val canSam == // arities and types line up, etc
  if (canSam) mkSam
  else withDiagnostic("Note: the function was not compatible with the SAM type ..") {
     mkFunction
  }
else
  mkFunction
```

This will be tricky, because `mkFunction` might work out fine, but we will be subsequently unable to adapt the function type to the almost-SAM. Perhaps we'll need a `-Xlog-sams` analagous to `-Xlog-implicits` to help diagnose why things didn't line up.

adriaanm committed:
Remove stray debug comment

adriaanm committed:
Simplify logic regarding missing parameter types

When we're typing the result of eta-expansion and fail
because of missing parameter types, we try again.
Whether we're in that case doesn't change between arguments,
so hoist that out of the loop.

Eventually, we should get rid of the retry and just immediately
type check the un-eta-expanded method call -- if possible...

adriaanm committed:
Single Abstract Method support: java8 test

Inspired by test/files/run/t7398.scala and sammy_poly.
Added some notes to original tests.

Elaborating on that note: we don't yet desugar `f(a)` to `f.sam(a)`,
like we do for regular functions: `f(a)` becomes `f.apply(a)`.

It seems pleasingly symmetrical and is easy to implement,
but not sure it's a good idea...

adriaanm commented:
bonus commit: adriaanm/scala@f02899e1be

retronym commented:
LGTM. We can address the remaining issues in a followup.

As part of the review, I've donated a few more test cases: https://github.com/adriaanm/scala/pull/9

Perhaps we should consider a new facility to allow us to synthesize:

``` scala
// (a, b) => body
class anon extends T {
  def foo(a: EmptyTree, b: EmptyTree) = magicScope {
    body
  }
}
```

in which `magicScope` hides `foo` and other other members of `anon` while typechecking `body` and where the `EmptyTree`-s for parameter types reuse the mechanism for method parameter type inference that backs `-Yinfer-argument-types`.

gkossakowski commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -979,6 +979,18 @@ trait Types
+     * Thus, findMember requiring DEFERRED flags yields deferred members,
+     * while `findMember(excludedFlags = 0, requiredFlags = 0).filter(_.isDeferred)` may not (if there's a corresponding concrete member)
+     *
+     * Requirements take precedence over exclusions, so requiring and excluding DEFERRED will yield a DEFERRED member (if there is one).
```
After looking at the code I believe the predecessor of this implication is false. I think the way to find out is to write JUnit tests. Added to my personal TODO list.

gkossakowski commented:
Doesn't merge cleanly anymore. I'll fix the merge and open a new PR.
